mautic/mautic
# Fix PHP deprecation errors - Issue #13270

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ x ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #13270 



#### Description:

When visiting the login page, the logs could have these deprecation notices:

```
PHP Deprecation - fnmatch(): Passing null to parameter #2 ($filename) of type string is deprecated - in file /app/middlewares/CORSMiddleware.php - at line 107
PHP Deprecated:  fnmatch(): Passing null to parameter #2 ($filename) of type string is deprecated in /app/middlewares/CORSMiddleware.php on line 107
```

This PR implements a simple fix proposed by @J-Wick4, where the variable $origin (which is passed as parameter 2 mentioned in the error message) is checked whether it is null before calling fnmatch(). If $origin is null, false is returned, and the function continues as before. This fixes the error messages.



#### Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Go to the [login page](https://mautic.ddev.site/s/login).
3. Refresh the page, and check that the logs do not have the afore mentioned errors.

If testing locally with docker, you can keep eyes on the logs through Docker. 

If testing on GitPod through browser, you can check the 'Problems' tab to see if the deprecated notice is there (expected string, found null).

![Screenshot 2024-04-09 at 16 05 06](https://github.com/mautic/mautic/assets/121946942/835bc3c6-dde4-470b-ab01-e2a50ee38a54)

If testing with GitPod but not through browser, you can run `ddev logs` in the ide terminal to see the logs.



If you want to first see the error messages, and then see the effect of the solution:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))

2. Replace function getAllowOriginHeaderValue(Request $request) with the following version. It makes sure the $origin is null and that the fnmatch is called in order to replicate the situation people were having with the deprecation notices in logs. in CORSMiddleware.php:

```
    private function getAllowOriginHeaderValue(Request $request)
    {
        // Temporary conditions for testing issue #13270
        $origin = null;
        $this->restrictCORSDomains = true; // Ensure domain restriction is enforced
        $this->validCORSDomains = ['*']; // Assuming a wildcard for simplicity

        // If we're not restricting domains, set the header to the request origin
        if (!$this->restrictCORSDomains || in_array($origin, $this->validCORSDomains)) {
            $this->requestOriginIsValid = true;

            return $origin;
        }

        // Check the domains using shell wildcard patterns
        $validCorsDomainFilter = function ($validCorsDomain) use ($origin) {
            return fnmatch($validCorsDomain, $origin, FNM_CASEFOLD);
        };
        if (array_filter($this->validCORSDomains, $validCorsDomainFilter)) {
            $this->requestOriginIsValid = true;
            $this->corsHeaders['Vary']  = 'Origin';

            return $origin;
        }

        $this->requestOriginIsValid = false;

        return null;
    }
```
3. Add the previous test code, go to the log in page, and refresh the page. Now the error messages should show in the logs (Either the 'Problems' tab in the GitPod browser, `ddev logs` result in ide terminal or logs through Docker).

4. Then to see the error messages fixed with the new solution, add the following code, refresh the log in page and check that the deprecation notice no longer appears:
```
    private function getAllowOriginHeaderValue(Request $request)
    {
        // Temporary conditions for testing issue #13270
        $origin = null;
        $this->restrictCORSDomains = true; // Ensure domain restriction is enforced
        $this->validCORSDomains = ['*']; // Assuming a wildcard for simplicity

        // If we're not restricting domains, set the header to the request origin
        if (!$this->restrictCORSDomains || in_array($origin, $this->validCORSDomains)) {
            $this->requestOriginIsValid = true;

            return $origin;
        }

        // Check the domains using shell wildcard patterns
        $validCorsDomainFilter = function ($validCorsDomain) use ($origin) {
            if ($origin === null) {
                return false;
            }

            return fnmatch($validCorsDomain, $origin, FNM_CASEFOLD);
        };

        if (array_filter($this->validCORSDomains, $validCorsDomainFilter)) {
            $this->requestOriginIsValid = true;
            $this->corsHeaders['Vary']  = 'Origin';

            return $origin;
        }

        $this->requestOriginIsValid = false;

        return null;
    }
```
